subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/NormalizedPathFingerprintCompareStrategy.java

```diff
@@ -106,17 +107,30 @@ protected boolean doVisitChangesSince(
         for (Entry<String, FileSystemLocationFingerprint> entry : currentFingerprints.entrySet()) {
             String currentAbsolutePath = entry.getKey();
             FileSystemLocationFingerprint currentFingerprint = entry.getValue();
-            List<FilePathWithType> previousFilesForFingerprint = missingPreviousFiles.get(currentFingerprint);
-            if (previousFilesForFingerprint.isEmpty()) {
+            boolean wasAccountedFor = accountFor(missingPreviousFiles, currentAbsolutePath, currentFingerprint);
+            if (!wasAccountedFor) {
                 FileType fingerprintType = currentFingerprint.getType();
                 results.put(currentFingerprint.getNormalizedPath(), new FilePathWithType(currentAbsolutePath, fingerprintType));
-            } else {
-                previousFilesForFingerprint.remove(0);
```

Comment:
Instead of treating every collision as equal (i.e. remove any element from `previousFilesForFingerprint`), we remove the one that has a matching absolute path instead.
This introduces a few behavior changes:
* if absolute path doesn't match any item, instead of accounting for it (like it was before, i.e. `remove(0)`), we treat it as added file
* if it matches any absolute path, it is treated as modified, instead of a different file being marked as modified

Reply:
For normalized path, the absolute path cannot affect if there are any changes are not. The only thing we can adjust/fix is which files we report as changes if there are any.

For example, if I move the whole inputs to a different location while keeping the file names intact, then the task should be up-to-date. With your changes, the task would not be up-to-date any more.

Reply:
Thanks for the comments, let me add a test for that!

Reply:
There have been tests, you just changed them in this PR (`new` vs. `old` absolute path strings).

Reply:
yeah, but those had the same normalized path, which didn't make sense to me. Was that correct?

Reply:
It was correct I think. The idea of the tests was to change the absolute paths while the normalized paths stay the same.

Reply:
I'll change them to use a parent folder, instead of a file name change, ok?

Reply:
That could also work. The point of the `one-old` vs. `one-new` etc. names was to make sure that the absolute paths are not taken into account. Using names like `old/one` vs `new/one` could suggest that some parts of the absolute path matter for the test, which would be false. Not a big distinction, but since I don't see `old/one` adding anything, I would leave the existing notation in place.

Reply:
Can we currently get to a state when the normalized path of `one-old` will be `one`?
It really confused me at first, it's why I think it would be better to go with folder names, which reflect possible application states.

Reply:
If it would have helped you then I'm not against using paths. Please go ahead and make the change.